pubsub/gossipsub/partial-messages.md

```diff
@@ -0,0 +1,189 @@
+in contrast to a content-based message id, does not require the full message to
+compute. For example, in the Ethereum use case, this could simply be the hash of
+the signed block header.
+
+The `topicID` references the Gossipsub topic a message, and thus its parts,
+belong to.
+
+The `metadata` field is opaque application defined metadata associated with this
+request. This can be a bitmap, a list of ranges, or a bloom filter. The
+application generates this and consumes this.
+
+
+Nodes SHOULD assume a `partialIWANT `implies a `IDONTWANT `for the full message.
+
+### PartialIDONTWANT
+
+PartialIDONTWANT serves to cancel any pending PartialIWANTs.
+
+A PartialIDONTWANT without metadata serves to cancel all pending PartialIWANTs,
+and signals there is no part this peer wants.
+
+Implementations SHOULD NOT send a PartialIHAVE to a peer with parts that the
+peer has previously sent a PartialIDONTWANT for.
+
+### PartialIHAVE
+
+PartialIHave allows nodes to signal HAVE information before receiving all
+segments, unlocking the use of partialIWANT in more contexts.
+
+In the context of partial messages, it is more useful than IHAVE as it tells the
+peer the group ID. In contrast, a IHAVE only includes a message ID that is
+unique to the message. A receiving peer has no way to link an IHAVE's message ID
+with a group ID, without having the full message.
+
+Partial IHAVE messages can be used both in the context of lazy push, notifying
+peers about reception progress, and in the context of heartbeats, sending
+also Partial IHAVEs.
+
+The structure of PartialIHAVE is analogous to that of PartialIWANT.
+
+The metadata, as in Partial IWants, is application defined. It is some encoding
+that represents the parts the sender has.
+
+Implementations are free to select when to send an update to their peers based
+on signaling bandwidth tradeoff considerations.
+
+## Application Interface
+
+Message contents are application defined. Thus splitting a message must be
+application defined. Here are a list of operations an application is expected to
+provide to Gossipsub to enable partial message delivery.
+
+1. Split a full message into a group of partial messages.
+2. Given a group of partial messages, encode a request for the missing parts. (Used by `PartialIWANT`)
+3. Given the request above and a group of partial messages, return the relevant
+   parts of the message. (Used to fulfill a `PartialIWANT`)
+4. Given a group of partial messages, encode metadata for the parts present. (Used by `PartialIHAVE`)
+5. Validate a partial message.
+6. Encode and decode a group of partial messages.
+
+## Protobuf
+
+```protobuf
+syntax = "proto2";
+
+message PartialMessagesExtension {
+  optional PartialMessage message = 1;
+  optional PartialIWANT iwant = 2;
+  optional PartialIDONTWANT idontwant = 3;
+  optional PartialIHAVE ihave = 4;
+}
+
+message PartialMessage {
+  optional bytes topicID = 1;
+  optional bytes data = 2;
+}
```

Comment:
Would the partial message benefit some contextual metadata to indicate which parts are actually being delivered? Or do we expect the data to encode this (e.g. with null elements inside a list, if that's the shape)? Or does the `PartialMessage` refer to a single part?

Also: this schema only allows for one inflight partial request per peer per topic. I guess that's not a problem and we (Ethereum) are fine with that limitation?